### PR TITLE
Spike: simplify providers

### PR DESCRIPTION
**Pull request by agualis**
# Description

We currently use [Roots.ts](https://github.com/balancer-labs/frontend-v2/blob/develop/src/Root.ts) to provide/inject global providers so that they can be used by any component. This solution works but it makes testing setup challenging, as it is really difficult to instantiate a component under test with access those injected providers.
The main goal of those providers is providing access to global state and methods/computed-props to interact with that state. That can be also done with plain composables as described in [this post](https://vueschool.io/articles/vuejs-tutorials/state-management-with-composition-api/). The benefits would be: 
- The new code is easier to understand cause there's no need for provide/inject or for the recursive calls in Root.ts.
- The new code is easier to test because the components using the composables simply import and call them (the don't need a higher level component to provide anything for them).
- Some TS interfaces like https://github.com/balancer-labs/frontend-v2/blob/d0dde3375f97fd39ba97146dad2cf6d58bafb037/src/providers/tokens.provider.ts#L51 could be deleted because the response type would be automatically inferred by TS.
- The relationships between "provider" modules become more explicit which helps detecting and fixing potential cyclic dependency problems. 

For the sake of a clearer review I kept the original file structure. For example: 
- `tokens-provider.ts` defines the global state and contains a public method that I called `getTokensProvision` which exports _state + computed + methods_.
- `useTokens.ts` simply export the previous method as `useTokens()`.

In a final refactor, I would merge those 2 files into one. 

IMO the biggest advantage of this change is that it allows us to test components without mocking the providers (we could still do that if we want) because we just need to mock the api calls with MSW (most of them could be mocked once globally to be reused for all tests). That would align with testing library mindset: 

> The more your tests resemble the way your software is used, the more confidence they can give you.

2 examples that illustrate how much we would simplify testing: 

https://github.com/balancer-labs/frontend-v2/pull/2375/files#diff-7b7c438140e37f8fdb2a851b9e0645c2983b6b91687870ee2f9f3f13df9ab782L8-L14 

https://github.com/balancer-labs/frontend-v2/pull/2375/files#diff-854b90a15a0fd9c299c4d8264d5fe9749d0efb15870c3491beaee364c2a90b6c

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency changes
- [x] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## How should this be tested?

Please provide instructions so we can test. Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

## Visual context

Please provide any relevant visual context for UI changes or additions. This could be static screenshots or a loom screencast.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have requested at least 2 reviews (If the PR is significant enough, use best judgement here)
- [ ] I have commented my code where relevant, particularly in hard-to-understand areas
- [ ] If package-lock.json has changes, it was intentional.
- [ ] The base of this PR is `master` if hotfix, `develop` if not

### CODE REVIEW

File: src/providers/token-lists.provider.ts [hunk 1]
```diff
@@ -21,121 +12,101 @@ export interface TokenListsState {
...
+    activeTokenLists,
+    defaultTokenList,
+    balancerTokenLists,
+    approvedTokenLists,
+    vettedTokenList,
+    // methods
+    toggleTokenList,
+    isActiveList,
+  };
+}
```
Comment (garethfuller) [inline src/providers/token-lists.provider.ts:112]: Why does this need to be a function? Also the Provision part of `getTokensListProvision` doesn't really make sense. What about:
```suggestion
export const tokenListsProvider = {
    // state
    ...toRefs(state),
    // computed
    allTokenLists,
    activeTokenLists,
    defaultTokenList,
    balancerTokenLists,
    approvedTokenLists,
    vettedTokenList,
    // methods
    toggleTokenList,
    isActiveList,
}
```

**Review by garethfuller — CHANGES_REQUESTED** [review]
You removed Root.ts which was injecting these providers, where are they all instantiated now?

File: src/providers/token-lists.provider.ts (hunk 1 above)
Comment (agualis) [reply src/providers/token-lists.provider.ts:112]: Yeah, I kept the two original files (`foo.provider.ts` and `useFoo.ts`) to make the code review simpler but the idea would be deleting `foo.provider.ts` with a const as you suggest.

**Comment by agualis** [comment]
> You removed Root.ts which was injecting these providers, where are they all instantiated now?

If we refactor the "providers" into composables, we don't need provide/inject anymore. They are simply imported and used in setup with `useFoo()`.

In the description, I tried to explain the advantages of it, although, there is one big disadvantage:

while the global code outside the `useFoo` function is just called once, the code inside useFoo is run every `useFoo` call.  As you know, that can affect performance in certain situations (like when using `watch`) but, I think, that we will be able to extract most of the code to the global scope in most of the cases.

**Comment by agualis** [comment]
@garethfuller Note that, in the tokens.provider.ts, we could extract more global computed and functions to the global scope so that they are not run every time the composable is used but there is one restriction: 

Functions/computed using vue-query cannot be outside the useProvider* function (global scope) because they need to be executed inside setup method. Because of that, I think it is better to leave everything inside `useProvider*` and move things up (to the global scope) just when performance is heavily affected.

**Comment by garethfuller** [comment]
Closing in favour of #2576

**Pull request closed without merging**